README.md
# SQLAlchemyCsvWriter

[![Stable Version](https://img.shields.io/pypi/v/sqlalchemy_csv_writer?label=stable)](https://pypi.org/project/sqlalchemy-csv-writer/#history)
![Python Versions](https://img.shields.io/pypi/pyversions/sqlalchemy_csv_writer)
![Tests](https://github.com/github/docs/actions/workflows/test.yml/badge.svg?branch=main)
[![codecov](https://codecov.io/gh/raayu83/sqlalchemy-csv-writer/graph/badge.svg?token=TXRKRRADUH)](https://codecov.io/gh/raayu83/sqlalchemy-csv-writer)
[![Ruff](https://img.shields.io/endpoint?url=https://raw.githubusercontent.com/astral-sh/ruff/main/assets/badge/v2.json)](https://github.com/astral-sh/ruff)
[![Poetry](https://img.shields.io/endpoint?url=https://python-poetry.org/badge/v0.json)](https://python-poetry.org/)

SQLAlchemyCsvWriter is a thin wrapper around the python csw.writer function to make the process of exporting SQLAlchemy query results to csv simpler. 

It supports both synchronous as well as asynchronous query results, as well as streaming.

## Features
- Write SQLAlchemy query results to csv file with little boilerplate
- Optionally write header by auto generation or passing column names
- Supports same dialect and formatting parameters as [csv.writer](https://docs.python.org/3/library/csv.html#csv.writer)
- Supports synchronous, asynchronous and asynchronous streaming results
- Supports any models defined with DeclarativeBase

## Installation
- with pip: `pip install sqlalchemy-csv-writer`
- with poetry: `poetry add sqlalchemy-csv-writer`

## Usage

```python
from sqlalchemy_csv_writer import SQLAlchemyCsvWriter

with SQLAlchemyCsvWriter(
    "test.csv",
    header=True,
    field_formats={"column_1": "%.2f"},
    prefix_model_names=True,
    dialect="unix",
) as writer:
    writer.write_rows(results)  # pass results of SQLAlchemy query
```

For full example see examples directory (uses the [alchemical](https://github.com/miguelgrinberg/alchemical) SQLAlchemy wrapper).

## Development Notes
- Linting and formatting using ruff
- Testing using pytest
- Dependency management and release using poetry
- Documentation using mkdocs with mkdocs-material and other plugins, released to GitHub pages
- CI/CD using GitHub Actions
- Pull requests are welcome

CHANGELOG.md
# Release Notes 

## 0.4.2
- tweak documentation

## 0.4.1
- tweak documentation

## 0.4.0
- measure coverage and add to docs
- adjust additional init parameters to match official csv.writer documentation

## 0.3.1
- add mkdocs and release to GitHub pages

## 0.3.0
- add support for passing string path or pathlib Path as csvfile parameter
- add support to use as context manager

## 0.2.0
- improve method naming

## 0.1.1
- fix minimum version requirement in pyprojects.toml

## 0.1.0
- initial release

tests/model.py
from alchemical import Model
from sqlalchemy import String
from sqlalchemy.orm import Mapped, mapped_column


class User(Model):
    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str] = mapped_column(String(128))
    value: Mapped[float] = mapped_column()


tests/test_writer_file.py
from pathlib import Path

import pytest
from alchemical import Alchemical
from sqlalchemy import select

from sqlalchemy_csv_writer import SQLAlchemyCsvWriter
from tests.model import User


@pytest.fixture(scope="module")
def db():
    db = Alchemical("sqlite:///:memory:")
    db.drop_all()
    db.create_all()

    with db.begin() as session:
        for name in ["mary", "joe", "susan"]:
            session.add(User(name=name, value=12.31))
    return db


def test_write_with_string_path(db):
    with db.Session() as session:
        test_file = "test/test.csv"
        Path(test_file).unlink(missing_ok=True)

        results = session.execute(select(User.id, User.name, User.value)).all()

        with SQLAlchemyCsvWriter(
            test_file,
            dialect="unix",
        ) as writer:
            writer.write_rows(results)

        expected_result = """"id","name","value"
"1","mary","12.31"
"2","joe","12.31"
"3","susan","12.31"
"""

        with open(test_file) as f:
            assert "".join(f.readlines()) == expected_result

        Path(test_file).unlink(missing_ok=True)


def test_write_with_pathlib_path(db):
    with db.Session() as session:
        test_file = Path("test/test2.csv")
        test_file.unlink(missing_ok=True)

        results = session.execute(select(User)).all()

        with SQLAlchemyCsvWriter(
            test_file,
            dialect="unix",
        ) as writer:
            writer.write_rows(results)

        expected_result = """"id","name","value"
"1","mary","12.31"
"2","joe","12.31"
"3","susan","12.31"
"""

        with open(test_file) as f:
            assert "".join(f.readlines()) == expected_result

        test_file.unlink(missing_ok=True)


tests/test_writer_sync.py
from io import StringIO

import pytest
from alchemical import Alchemical
from sqlalchemy import select

from sqlalchemy_csv_writer import SQLAlchemyCsvWriter
from tests.model import User


@pytest.fixture(scope="module")
def db():
    db = Alchemical("sqlite:///:memory:")
    db.drop_all()
    db.create_all()

    with db.begin() as session:
        for name in ["mary", "joe", "susan"]:
            session.add(User(name=name, value=12.31))
    return db


def test_write_rows_with_scalars_with_prefixed_header_and_formatting(db):
    with db.Session() as session:
        stringio = StringIO()
        results = session.scalars(select(User)).all()

        field_formats = {"User.value": "%.1f"}
        writer = SQLAlchemyCsvWriter(
            stringio,
            field_formats=field_formats,
            prefix_model_names=True,
            dialect="unix",
        )
        writer.write_rows(results)

        expected_result = """"User.id","User.name","User.value"
"1","mary","12.3"
"2","joe","12.3"
"3","susan","12.3"
"""

        assert stringio.getvalue() == expected_result


def test_write_rows_with_execute_with_header(db):
    with db.Session() as session:
        stringio = StringIO()
        results = session.execute(select(User)).all()

        writer = SQLAlchemyCsvWriter(
            stringio,
            dialect="unix",
        )
        writer.write_rows(results)

        expected_result = """"id","name","value"
"1","mary","12.31"
"2","joe","12.31"
"3","susan","12.31"
"""

        assert stringio.getvalue() == expected_result


def test_write_rows_with_execute_without_header(db):
    with db.Session() as session:
        stringio = StringIO()
        results = session.execute(select(User)).all()

        writer = SQLAlchemyCsvWriter(
            stringio,
            header=False,
            dialect="unix",
        )
        writer.write_rows(results)

        expected_result = """"1","mary","12.31"
"2","joe","12.31"
"3","susan","12.31"
"""

        assert stringio.getvalue() == expected_result


def test_write_rows_with_execute_with_duplicate_columns(db):
    with db.Session() as session:
        stringio = StringIO()
        results = session.execute(select(User, User)).all()

        writer = SQLAlchemyCsvWriter(
            stringio,
            header=True,
            dialect="unix",
        )
        writer.write_rows(results)

        expected_result = """"id","name","value","id","name","value"
"1","mary","12.31","1","mary","12.31"
"2","joe","12.31","2","joe","12.31"
"3","susan","12.31","3","susan","12.31"
"""

        assert stringio.getvalue() == expected_result


def test_write_rows_with_custom_header(db):
    with db.Session() as session:
        stringio = StringIO()
        results = session.execute(select(User)).all()

        writer = SQLAlchemyCsvWriter(
            stringio,
            header=["col_1", "col_2", "col_3"],
            dialect="unix",
        )
        writer.write_rows(results)

        expected_result = """"col_1","col_2","col_3"
"1","mary","12.31"
"2","joe","12.31"
"3","susan","12.31"
"""

        assert stringio.getvalue() == expected_result


def test_write_rows_with_custom_header_error(db):
    with db.Session() as session:
        stringio = StringIO()
        results = session.execute(select(User)).all()

        writer = SQLAlchemyCsvWriter(
            stringio,
            header=["col_1", "col_2"],
            dialect="unix",
        )

        with pytest.raises(ValueError):
            writer.write_rows(results)


examples/example_sync.py
from pathlib import Path

from alchemical import Alchemical, Model
from sqlalchemy import String, select
from sqlalchemy.orm import Mapped, mapped_column

from sqlalchemy_csv_writer import SQLAlchemyCsvWriter


class User(Model):
    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str] = mapped_column(String(128))
    value: Mapped[float] = mapped_column()


db = Alchemical("sqlite:///:memory:")
db.drop_all()
db.create_all()

with db.begin() as session:
    for name in ["mary", "joe", "susan"]:
        session.add(User(name=name, value=12.3))

with db.Session() as session:
    results = session.execute(select(User)).all()

    field_formats = {"value": "%.2f"}
    with SQLAlchemyCsvWriter(
        Path("test/example_sync.csv"),
        header=True,
        field_formats=field_formats,
        prefix_model_names=True,
        dialect="unix",
    ) as writer:
        writer.write_rows(results)


examples/example_async.py
import asyncio
from io import StringIO

from alchemical.aio import Alchemical, Model
from sqlalchemy import String, select
from sqlalchemy.orm import Mapped, mapped_column

from sqlalchemy_csv_writer import SQLAlchemyCsvWriter


class User(Model):
    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str] = mapped_column(String(128))
    value: Mapped[float] = mapped_column()


async def run():
    db = Alchemical("sqlite:///:memory:")
    await db.drop_all()
    await db.create_all()

    async with db.begin() as session:
        for name in ["mary", "joe", "susan"]:
            session.add(User(name=name, value=12.3))

    async with db.Session() as session:
        stringio = StringIO()
        results = await session.execute(select(User))

        field_formats = {"value": "%.2f"}
        writer = SQLAlchemyCsvWriter(
            stringio,
            header=True,
            field_formats=field_formats,
            dialect="unix",
        )
        writer.write_rows(results)

        print(stringio.getvalue())


asyncio.run(run())


sqlalchemy_csv_writer/writer.py
"""Provide SQLAlchemy Csv Writer."""
import csv
import typing
from pathlib import Path
from typing import Union

from sqlalchemy import inspect
from sqlalchemy.orm import DeclarativeBase


class SQLAlchemyCsvWriter:
    """Write SQL Alchemy results to a csv file."""

    def __init__(
        self,
        csvfile: Union[Path, str, typing.IO],
        header: Union[list[str], bool] = True,
        prefix_model_names: bool = False,
        field_formats: Union[dict[str, str], None] = None,
        dialect="excel",
        **fmtparams,
    ):
        """Create a SQLAlchemyCsvWriter instance.

        The instance's methods can be used to write rows to the specified csv file.

        Args:
            csvfile: Path or File-like object to write the resulting csv data to
            header: True to automatically generate header, False to disable header or list of strings for custom header
            prefix_model_names:  Whether to prefix the model names in the header
            field_formats: Dictionary containing the column name as keys and column format as a values (using % style format syntax)
            dialect: csv dialect to use
            **fmtparams: extra formatting parameters to pass to csv.writer instance
        """
        if isinstance(csvfile, str):
            csvfile = Path(csvfile)
        if isinstance(csvfile, Path):
            csvfile.parent.mkdir(exist_ok=True, parents=True)
            csvfile = open(csvfile, "w", encoding="utf-8")  # noqa: SIM115

        self.csvfile = csvfile
        self.writer = csv.writer(csvfile, dialect=dialect, **fmtparams)
        self.header = header
        self.prefix_model_names = prefix_model_names
        self.field_formats = field_formats if field_formats else {}
        self.header_row_written = False

    def __del__(self):
        """Close open resources."""
        if hasattr(self.csvfile, "close"):
            self.csvfile.close()

    def __enter__(self):
        """SQLAlchemyCsvWriter may be used as a context manager."""
        return self

    def __exit__(self, type, value, traceback):
        """Exit context manager."""
        self.__del__()

    async def write_rows_stream(self, results: list):
        """Write query results to csv.

        Write query results retrieved with SQLAlchemy's .stream or .stream_scalars.

        Args:
            results: query results retrieved with SQLAlchemy's .stream or .stream_scalars
        """
        async for result in results:
            self._process_result(result)

    def write_rows(self, results: list):
        """Write query results to csv.

        Write query results retrieved with SQLAlchemy's .execute or .scalars. to csv

        Args:
            results: query results retrieved with SQLAlchemy's .execute or .scalars
        """
        for result in results:
            self._process_result(result)

    def _process_result(self, result):
        result = self._extract_columns(result)

        # write header
        if self.header and not self.header_row_written:
            if self.header is True:
                self.writer.writerow([r[0] for r in result])
            else:
                if len(self.header) == len(result):
                    self.writer.writerow(self.header)
                else:
                    raise ValueError("Length of header and content does not match.")
            self.header_row_written = True

        # write data
        values = []
        for key, value in result:
            if str(key) in self.field_formats:
                value = self.field_formats[str(key)] % value
            values.append(value)
        self.writer.writerow(values)

    def _extract_columns(self, result):
        columns = []
        if hasattr(result, "_mapping"):  # is not a scalar
            for element_key, element_value in result._mapping.items():
                if isinstance(element_value, DeclarativeBase):  # is an orm model
                    columns.extend(self._extract_model(element_value))
                else:  # is a column
                    columns.append((element_key, element_value))
        elif isinstance(result, DeclarativeBase):  # is a scalar orm model
            columns.extend(self._extract_model(result))

        return columns

    def _extract_model(self, obj):
        insp = inspect(obj)
        if self.prefix_model_names:
            return [(getattr(insp.class_, attr.key), attr.value) for attr in insp.attrs]
        else:
            return [(attr.key, attr.value) for attr in insp.attrs]


examples/example_async_stream.py
import asyncio
from io import StringIO

from alchemical.aio import Alchemical, Model
from sqlalchemy import String, select
from sqlalchemy.orm import Mapped, mapped_column

from sqlalchemy_csv_writer import SQLAlchemyCsvWriter


class User(Model):
    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str] = mapped_column(String(128))
    value: Mapped[float] = mapped_column()


async def run():
    db = Alchemical("sqlite:///:memory:")
    await db.drop_all()
    await db.create_all()

    async with db.begin() as session:
        for name in ["mary", "joe", "susan"]:
            session.add(User(name=name, value=12.3))

    async with db.Session() as session:
        stringio = StringIO()
        results = await session.stream(select(User))

        field_formats = {"value": "%.2f"}
        writer = SQLAlchemyCsvWriter(
            stringio,
            header=True,
            field_formats=field_formats,
            dialect="unix",
        )
        await writer.write_rows_stream(results)

        print(stringio.getvalue())


asyncio.run(run())
